File: program19.c
#include <stdio.h>

#define INF 9999
#define MAX 20

int main() {
    int n, i, j, k;
    int dist[MAX][MAX];

    printf("Enter number of vertices: ");
    scanf("%d", &n);

    printf("Enter adjacency matrix (0 for no edge):\n");
    for (i = 0; i < n; i++)
        for (j = 0; j < n; j++) {
            scanf("%d", &dist[i][j]);
            if (i != j && dist[i][j] == 0)
                dist[i][j] = INF;
        }

    // Floyd–Warshall Algorithm
    for (k = 0; k < n; k++)
        for (i = 0; i < n; i++)
            for (j = 0; j < n; j++)
                if (dist[i][j] > dist[i][k] + dist[k][j])
                    dist[i][j] = dist[i][k] + dist[k][j];

    printf("\nShortest Distance Matrix:\n");
    for (i = 0; i < n; i++) {
        for (j = 0; j < n; j++) {
            if (dist[i][j] == INF)
                printf("INF ");
            else
                printf("%3d ", dist[i][j]);
        }
        printf("\n");
    }

    return 0;
}

File: program19.py
import sys

def dijkstra(graph, start):
    V = len(graph)
    distance = [sys.maxsize] * V
    visited = [False] * V

    distance[start] = 0

    for _ in range(V):
        # Find vertex with minimum distance
        min_dist = sys.maxsize
        min_index = -1

        for v in range(V):
            if not visited[v] and distance[v] < min_dist:
                min_dist = distance[v]
                min_index = v

        visited[min_index] = True

        # Update distance values
        for v in range(V):
            if graph[min_index][v] != 0 and not visited[v]:
                new_dist = distance[min_index] + graph[min_index][v]
                if new_dist < distance[v]:
                    distance[v] = new_dist

    return distance


# -------- MAIN PROGRAM --------
V = int(input("Enter number of vertices: "))
graph = []

print("\nEnter adjacency matrix (use 0 where no edge exists):")
for i in range(V):
    row = list(map(int, input().split()))
    graph.append(row)

start = int(input("Enter starting vertex (0 to V-1): "))
result = dijkstra(graph, start)

print("\nShortest distances from vertex", start)
for i in range(V):
    print(f"{start} → {i} = {result[i]}")
